src/Utilities/Math.hpp

```diff
@@ -162,3 +164,31 @@ template <typename T>
 constexpr T sgn(const T& val) {
   return sgn_detail::sgn(val, std::is_signed<T>{});
 }
+
+/// \ingroup UtilitiesGroup
+/// \brief Raises type T to the integer power n.
+///
+/// \details The power is computed by squaring T to the closest power of 2 lower
+/// than n and then multiplying the remainder on. There are implementations with
+/// fewer multiplications but those require at least one allocation which is
+/// avoided here.
```

Comment:
SpEC has the algorithm you want.

Reply:
the SpEC algorithm also allocates. Are we sure we want that? We have often said that we can do a *a lot* of extra calculations before they become as expensive as an allocation.

Reply:
Where do you see temporaries?
```c++
  template<class T> T exp_by_squaring(const T b, const int e) {
    int ecount = e;
    int bitcount = 1;
    while(ecount >>= 1) {
      ++bitcount;
    }
    T ret = 1;
    while(bitcount) {
      ret *= ret;
      if((e >> --bitcount) & 0x1) {
        ret *= b;
      }
    }
    return ret;
  }
```

Reply:
ah sorry, I was looking at the wrong function